feat(ValidHookName): Add sniff for hook name in PHP attributes

RuZniki opened the pull request:
Issue: https://www.drupal.org/project/coder/issues/3499291

klausi commented on coder_sniffer/Drupal/Sniffs/NamingConventions/ValidAttributeHookNameSniff.php:
```diff
@@ -0,0 +1,69 @@
+ * @link     http://pear.php.net/package/PHP_CodeSniffer
+ */
+
+namespace Drupal\Sniffs\NamingConventions;
```
In the naming conventions folder we have PHP names, so I don't think the property string of an attribute belongs there.

Let's make a new folder "Attributes" and just call the class "ValidHookNameSniff"

klausi commented on coder_sniffer/Drupal/Sniffs/NamingConventions/ValidAttributeHookNameSniff.php:
```diff
@@ -0,0 +1,69 @@
+    {
+        $tokens = $phpcsFile->getTokens();
+
+        if ($tokens[$stackPtr + 1]['type'] === 'T_STRING'
```
for all those conditions we should check first with isset() if the next token is there. This will help prevent failures in live coding when the code is just being written.

RuZniki replied:
I swtiched to `->findNext`, do you think this is enough?

klausi commented on coder_sniffer/Drupal/Sniffs/NamingConventions/ValidAttributeHookNameSniff.php:
```diff
@@ -0,0 +1,69 @@
+        if ($tokens[$stackPtr + 1]['type'] === 'T_STRING'
+            && $tokens[$stackPtr + 1]['content'] === 'Hook'
+            && $tokens[$stackPtr + 3]['type'] === 'T_CONSTANT_ENCAPSED_STRING'
+            && str_contains($tokens[$stackPtr + 3]['content'], 'hook_')
```
we need to make an exception for `hook_info`.

RuZniki replied:
[nicxvan](https://www.drupal.org/u/nicxvan) mentioned that ([link](https://www.drupal.org/project/coder/issues/3499291#comment-15947940)) hook_hook_info cannot be oop, but I decide to add it anyway. Maybe it will be more exceptions in the future.

klausi commented on coder_sniffer/Drupal/Sniffs/NamingConventions/ValidAttributeHookNameSniff.php:
```diff
@@ -0,0 +1,69 @@
+        ) {
+            $hookName = $tokens[$stackPtr + 3]['content'];
+            $phpcsFile->addFixableWarning('Hook name should not start with "hook_" prefix. Hook name used:' . $hookName, $stackPtr + 3,'AttributePrefixHookName');
+            $phpcsFile->fixer->replaceToken($stackPtr + 3, str_replace('hook_', '', $hookName));
```
before replacing tokens we should check if the fixer is active, by assigning a `$fix` variable on the line before and checking that. See how we do that in other places.

klausi commented on tests/Drupal/bad/BadUnitTest.php:
```diff
@@ -384,7 +384,7 @@ protected function getErrorList(string $testFile): array
                 860 => 2,
                 867 => 1,
-                871 => 2,
+                879 => 2,
```
Since this sniff will likely not interfere with others we should leave BadUnitTest alone and do this in a dedicated test file for this new sniff.

klausi commented on tests/Drupal/bad/bad.php:
```diff
@@ -868,4 +868,12 @@ function test30(TestType $a = NULL) {
+ * Implements hook_node_view().
+ */
+#[Hook('hook_node_view')]
+function bad_node_view() {
```
please also add a test for `hook_info` that is passes in the new sniff test file.

klausi reviewed (changes requested):
@RuZniki nice to hear from you :)

Left some comments, please also check the test fails.

RuZniki commented on coder_sniffer/Drupal/Sniffs/Attributes/ValidHookNameSniff.php:
```diff
@@ -0,0 +1,90 @@
+     *
+     * @var string[]
+     */
+    public array $hookExceptions = ['hook_info'];
```
Not sure, but typed properties came at php 7.4, in composer `"php": ">=7.2"`. Should I remove type property?

RuZniki commented:
@klausi I am also happy to get feedback from you, thanks a lot for your review.

I moved sniff to `\Drupal\Sniffs\Attributes\ValidHookNameSniff`.
And moved tests from bad.php to `\Drupal\Test\Attributes\ValidHookNameUnitTest`.
I decide not to test, when name arguments for Hook attribute in wrong order. Example:
```php
#[Hook(method: "userLogin", hook: "hook_user_login")] 
```
I believe this should not be a part of this sniff.

I add special handling for hook_hook_info(). But still not sure about this case.

PR is ready for review.

RuZniki commented:
I removed special handling for "hook_hook_info()".

klausi commented on coder_sniffer/Drupal/Sniffs/Attributes/ValidHookNameSniff.php:
```diff
@@ -0,0 +1,81 @@
+
+                if (strpos($hookNameValue, 'hook_') === 0) {
+                    $fix = $phpcsFile->addFixableWarning("Hook name should not start with 'hook_'. Hook name used: $hookNameValue", $hookName, 'HookPrefix');
+                    if ($fix === true && strlen($hookNameValue) > 5) {
```
the strlen() condition here can be removed, this was already checked before.

We should not add additional conditions with `$fix`, it would be very confusing if the fixer does not always work.

RuZniki replied:
I tried to avoid case when hook is not finished, and it's just "hook_".
Now I get your point, I will move second clause to above statement. Thanks.

klausi commented on coder_sniffer/Drupal/Sniffs/Attributes/ValidHookNameSniff.php:
```diff
@@ -0,0 +1,81 @@
+                    $fix = $phpcsFile->addFixableWarning("Hook name should not start with 'hook_'. Hook name used: $hookNameValue", $hookName, 'HookPrefix');
+                    if ($fix === true && strlen($hookNameValue) > 5) {
+                        // Remove "hook_" prefix.
+                        $hookNameValueFixed = substr($hookNameValue, 5);
```
we should use this in the error message above like "The hook name should not start with 'hook_', expected '%s' but found '%s".

RuZniki replied:
Thanks, I update error message.
Result example

> The hook name should not start with 'hook_', expected 'node_alter' but found 'hook_node_alter' (Drupal.Attributes.ValidHookName.HookPrefix)

klausi reviewed (changes requested):
Sorry for the delay!

Looks good to me, just 2 minor things inline.

klausi commented:
Thanks, Drupal core regression tests are failing, but looks unrelated to this issue. Will check in a follow-up!
